### PR TITLE
WIP: Implementation of cross frequency coupling. 

### PR DESCRIPTION
**Pull request by pravsripad**
Includes some functions to compute the cross frequency coupling and the modulation index as described in: - 

1. High gamma power is phase-locked to theta oscillations in human neocortex, Canolty 2006. (http://www.ncbi.nlm.nih.gov/pubmed/16973878)
 
2. Measuring phase-amplitude coupling between neuronal oscillations of different frequencies, Tort, 2010.
(http://www.ncbi.nlm.nih.gov/pubmed/20463205)

Things to be done: - 

1. Extensive testing of implementation with reliable sample data. 
2. Improve API, make more mne (structure, documentation, logging etc.)
3. Decide how to integrate / use with raw and epochs. 

Any suggestions/ideas on how best to do this?

### CODE REVIEW

**Comment by dengemann** [comment]
coucou @pravsripad, where are we here? Any chances we can make this mergeable soonish?

**Comment by pravsripad** [comment]
@dengemann sorry, have put this on hold for quite sometime now. Will give it a go again this weekend, are we approaching a release soon?

**Comment by dengemann** [comment]
Yes, we're releasing soon.

2015-04-17 9:05 GMT+02:00 Praveen Sripad <notifications@github.com>:

**Comment by dengemann** [comment]
@pravsripad ping ping. This is almost 1 year in the air ...

File: mne/connectivity/cfc.py [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+        Data to be shuffled.
+
+    Returns
+    -------
+    shuffled_data : ndarray, shape (n_epochs, n_times)
+        The shuffled data, with each epoch shuffled across n_times.
+    """
+    shuffled_data = data.copy()
+    map(shuffle, shuffled_data)
+    return shuffled_data
```
Comment (dengemann) [inline mne/connectivity/cfc.py]: we should use phase-scrambling here. Ideally picking one scrambling algo after some literature review

Comment (pravsripad) [inline mne/connectivity/cfc.py]: @dengemann do you mean randomizing the phases and not just the time points? Something like this? 

```

    from sklearn.utils import check_random_state
    data = np.asarray(data)
    data_freq = np.fft.rfft(data, axis=0)
    rng = check_random_state(random_state)
    data_freq = np.abs(data_freq) * np.exp(1j * rng.random_sample(data_freq.shape) * 2 * np.pi)
    return np.fft.irfft(data_freq, data.shape[0], axis=0)
```

Comment (dengemann) [inline mne/connectivity/cfc.py]: Yes, something like that. Also see here for an overview:

http://en.wikipedia.org/wiki/Surrogate_data_testing

Scanning the EEG/MEG literature a bit it seems the phase scrambling seems pretty common.
The nice thing is, that it only touches the phase of the signal.

Comment (pravsripad) [inline mne/connectivity/cfc.py]: Yes it is, I've used it before. Will replace the surrogates function with it then.

Comment (dengemann) [inline mne/connectivity/cfc.py]: Did you see differences?

Comment (pravsripad) [inline mne/connectivity/cfc.py]: For connectivity analysis where I have used this, I would say there is not much of a visible difference in results, but theoretically and in literature randomizing only the phase is used and it makes sense. 

surr_pochs

Comment (dengemann) [inline mne/connectivity/cfc.py]: I think it depends a bit on the null-hypothesis. To me the phase-scrambling seems more precise, as ofr these kind of analyses you have hypotheses regarding the phase, etc.

**Comment by choldgraf** [comment]
Ah I see this is already well underway. I think that the methods you guys have worked out in there are more thoroughly developed, though less diverse. My guess is that (at least for a first round) you guys would want to get this code implemented and merged before adding any extra CFC computation methods, right? (e.g. CFC by linear model).

Quick thought - it looks like these are all implemented to be done within a single signal, is that correct? Any plans to allow for calculating PAC / CFC between two signals? E.g., include a `data_high` flag that would allow you to supply a signal for the high-frequency component, and if it is `none` then just default to the current behavior. Alternatively supply data in the form of `shape(n_epochs, n_chans, n_times)` and then specify a connectivity array the same way that the other connectivity functions work?

**Comment by agramfort** [comment]
everything is doable :)

**Comment by choldgraf** [comment]
Haha, but is it "doable in this PR"?

**Comment by agramfort** [comment]
this PR is dead sadly...

**Comment by agramfort** [comment]
or needs love I should say...

**Comment by choldgraf** [comment]
haha, don't we all. ok well I can put some time into this next week w/ folks if that'd be helpful.

**Comment by choldgraf** [comment]
What is left to be improved on this PR? It seems to be quite close, no? Alternatively is it worth writing out a roadmap for any other changes / enhancements to make?

**Comment by agramfort** [comment]
can you do a review of what's left to do for you?

**Comment by choldgraf** [comment]
ya, just wanted to see if this already existed in case someone else had already done it

**Comment by pravsripad** [comment]
Hello, sorry to join into the discussion late and I am sorry I have not been able to put more time into this PR. It is nearly done except that the unit tests are not extensive, the coverage does not exceed 90% yet. 

@choldgraf yes, in this cfc, pac is implemented within one signal only. I just pushed the last of my changes.

**Comment by pravsripad** [comment]
@agramfort @choldgraf The last of the above comments have been implemented as well. The functions for stats and plotting are decoupled and cosmetic changes suggested above have been added. Only improving the test coverage remains. 

We can discuss how to take this further.

**Comment by choldgraf** [comment]
Ok cool - is it close enough to being ready that it's better to just review this and pull it soon, or worth it to consider making changes before then? Would it be much work to include an option to provide a separate signal (e.g. `signal_high`) and for now make it default to "None"? I think at some point someone will want to do CFC or PAC between signals rather than within a signal, no?

**Comment by Eric89GXL** [comment]
@choldgraf it is my sense that you should take over the PR. That way you
can implement whatever you all settle on. Do you know how to take it over?
I can describe if not.

**Comment by pravsripad** [comment]
@choldgraf  To modify the code to pass a single signal would involve quite a few changes. It is currently suited to compute CFC/PAC between different frequency bands only. I am not sure of the use case to compute CFC between different signals, but I think it is a good idea since it is more a generic approach. 

+1 from me if you would like to take over this PR.

**Comment by pravsripad** [comment]
There are currently some conflicts, is a rebase required from my end?

**Comment by Eric89GXL** [comment]
@pravsripad yes either you or @choldgraf will need to rebase, up to you

**Comment by choldgraf** [comment]
If you rebase then I am happy to take over from there. @Eric89GXL does it basically just entail forking his fork, and then setting the upstream to the MNE master branch?

**Comment by Eric89GXL** [comment]
No, forking means "make a copy of code on GitHub", which you don't need to do. What you need to do is make a branch (on your local machine) that looks like his (out on github). The way you make local branches look like remote ones is by populating entries in your `git remote`. For example, on my machine:
```Bash
$ git remote -v
larsoner@bunk:~/custombuilds/mne-python$ git remote -v
adykstra	git://github.com/adykstra/mne-python.git (fetch)
adykstra	git://github.com/adykstra/mne-python.git (push)
agramfort	git@github.com:/agramfort/mne-python.git (fetch)
agramfort	git@github.com:/agramfort/mne-python.git (push)
...
choldgraf	git://github.com/choldgraf/mne-python.git (fetch)
choldgraf	git://github.com/choldgraf/mne-python.git (push)
...
origin	git@github.com:Eric89GXL/mne-python.git (fetch)
origin	git@github.com:Eric89GXL/mne-python.git (push)
...
upstream	git@github.com:mne-tools/mne-python.git (fetch)
upstream	git@github.com:mne-tools/mne-python.git (push)
...
```
This is because I have added entries for folks whose branches I've needed to work on, test, etc. So you should be able to do:
```Bash
$ git remote add pravsripad git://github.com/pravsripad/mne-python.git
$ git fetch pravsripad
$ git checkout -b cross-freq pravsripad/cross_freq_coupling
```
This will 1) add an entry in your `remote` list, so your computer knows about his fork, 2) goes out and finds all branches/changes in his fork, and 3) creates and checks out a new branch on your local machine named `cross-freq` that looks like (and tracks) his remote one.

**Comment by choldgraf** [comment]
So then in that case I would have push access to @pravsripad's branch, and this PR would update automatically as I pushed things?

**Comment by Eric89GXL** [comment]
It would if you had push access to his repo. In most cases you just close the PR and open a new one so that everyone doesn't have to give out write access, though.

**Comment by choldgraf** [comment]
ah ok, so here’s what I’d do:

   1. He rebases to master and I rebase to master (just to make sure we're
   on the same page)
   2. I pull his branch to my MNE repo w/ the instructions you just said
   3. I push this branch to my own MNE repo on github
   4. I open a PR with my branch
   5. Further changes are made to my branch and we go from there

That sounds correct?
​

**Comment by Eric89GXL** [comment]
Yes, except there should only ever be one rebase. Either you wait until he rebases, then check out his branch, or you check out his branch now, and then rebase it yourself.

**Comment by choldgraf** [comment]
ah good point - once @pravsripad is rebased / up to date then I will pull the branch to mine.

**Comment by pravsripad** [comment]
@choldgraf the branch is rebased and up to date.. If all the checks pass, it should be ready to pull.

**Comment by TomDLT** [comment]
:facepunch:  Git-Fu Tip:  A faster way to copy a PR in your own computer is to use : 
`git fetch upstream pull/1388/head:my_copy_of_pr_1388`
(where `upstream` is the usual name of the main remote)

Then you can easily test it, rebase it, push it to your github fork, ...
It also helps keeping your remote list clean.

**Comment by Eric89GXL** [comment]
What's wrong with having 100 entries in your remote? :)

**Comment by mainakjas** [comment]
welcome to mne-python @TomDLT :)

File: mne/connectivity/cfc.py (hunk 1 above)
Comment (choldgraf) [inline mne/connectivity/cfc.py]: @pravsripad did you implement code to do phase shuffling? I don't see it in the code.

Comment (dengemann) [inline mne/connectivity/cfc.py]: I have it locally, can contribute it.

Comment (choldgraf) [inline mne/connectivity/cfc.py]: I also found this:

http://www.pik-potsdam.de/~donges/pyunicorn/_modules/pyunicorn/timeseries/surrogates.html#Surrogates.correlated_noise_surrogates

On Thu, Jul 16, 2015 at 7:09 PM, Denis A. Engemann <notifications@github.com

Comment (choldgraf) [inline mne/connectivity/cfc.py]: actually look to be a lot of good surrogate functions in there

File: mne/connectivity/cfc.py [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+
+    # Calculate phase series of phase modulating signal
+    phase_series_fp = np.angle(hilbert(x_fp)) + np.pi
+
+    # Calculate amplitude envelope of amplitude modulated signal
+    amp_envelope_fa = np.abs(hilbert(x_fa))
+
+    if surrogates is True:
+        phase_series_fp = shuffle_data(phase_series_fp)
+        amp_envelope_fa = shuffle_data(amp_envelope_fa)
```
Comment (pravsripad) [inline mne/connectivity/cfc.py]: @dengemann @choldgraf I didn't implement the phase scrambling code snippet mentioned above. This is because of the way I use shuffle_data directly on the phases here.

File: mne/connectivity/cfc.py (hunk 1 above)
Comment (agramfort) [inline mne/connectivity/cfc.py]: license is BSD so we're good if you want to reuse. Just try to focus on the
minimal viable product here.

Comment (dengemann) [inline mne/connectivity/cfc.py]: I convinced them to use BSD, it's my friend Jonathan ;) Lot's of SWIG code, not sure we want this. We can get some inspiration though, and locally I have a working phase-scrambling.

**Comment by Eric89GXL** [comment]
@choldgraf has taken over in another PR, closing
